File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/tools/LiteBashTool.java
package com.zzf.rikki.idea.agent.tools;

import com.fasterxml.jackson.databind.JsonNode;
import com.zzf.rikki.idea.agent.compat.CommandRunner;
import com.zzf.rikki.idea.agent.compat.CommandRunnerResult;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

public class LiteBashTool {
    public static final Companion Companion = new Companion();
    private final CommandRunner runner = new CommandRunner();

    public String execute(JsonNode args, String workspaceRoot, AtomicBoolean skipFlag) {
        CommandRunnerResult detail = executeDetailed(args, workspaceRoot, skipFlag);
        long timeoutMs = args.path("timeout").asLong(CommandRunner.DEFAULT_TIMEOUT_MS);
        if (timeoutMs <= 0L) {
            timeoutMs = CommandRunner.DEFAULT_TIMEOUT_MS;
        }
        return formatResult(args.path("command").asText(""), timeoutMs, detail);
    }

    public CommandRunnerResult executeDetailed(JsonNode args, String workspaceRoot, AtomicBoolean skipFlag) {
        String command = args.path("command").asText("");
        if (command.isBlank()) {
            throw new IllegalArgumentException("command is required");
        }
        String workdir = args.path("workdir").asText(workspaceRoot);
        if (workdir.isBlank()) {
            workdir = workspaceRoot;
        }
        long timeoutMs = args.path("timeout").asLong(CommandRunner.DEFAULT_TIMEOUT_MS);
        if (timeoutMs <= 0L) {
            timeoutMs = CommandRunner.DEFAULT_TIMEOUT_MS;
        }
        String shell = args.path("shell").asText("auto").trim().toLowerCase(Locale.ROOT);
        if (shell.isBlank()) {
            shell = "auto";
        }
        return runner.run(command, workspaceRoot, workdir, timeoutMs, shell, skipFlag);
    }

    public String formatResult(String command, long timeoutMs, CommandRunnerResult result) {
        return runner.formatOutput(command, timeoutMs, result);
    }

    public static String commandFamily(String command) {
        return Companion.commandFamily(command);
    }

    public static final class Companion {
        private static final List<Map.Entry<String, String>> RISK_PATTERNS = List.of(
                Map.entry("sudo ", "requires elevated privileges"),
                Map.entry("su -", "switches user identity"),
                Map.entry("su root", "switches user identity"),
                Map.entry("rm -rf", "destructive file removal"),
                Map.entry("rm -fr", "destructive file removal"),
                Map.entry("rm -r ", "recursive deletion"),
                Map.entry("rm -f /", "targets filesystem root"),
                Map.entry("mkfs", "formats a filesystem"),
                Map.entry("dd if=", "raw disk write"),
                Map.entry("| bash", "pipes remote script into shell"),
                Map.entry("| sh ", "pipes remote script into shell"),
                Map.entry("| zsh ", "pipes remote script into shell"),
                Map.entry("| fish ", "pipes remote script into shell"),
                Map.entry("chmod 777", "grants unsafe world-writable permissions"),
                Map.entry("chmod -R ", "recursively changes permissions"),
                Map.entry("> /dev/", "writes to device path"),
                Map.entry("/dev/sd", "targets block device"),
                Map.entry("/dev/hd", "targets block device"),
                Map.entry("/dev/nvme", "targets block device"),
                Map.entry(":(){ :|:& };:", "fork bomb")
        );

        public boolean isHighRiskCommand(String command) {
            return !detectRiskReasons(command).isEmpty();
        }

        public List<String> detectRiskReasons(String command) {
            String normalized = command == null ? "" : command.trim().toLowerCase(Locale.ROOT);
            List<String> reasons = new ArrayList<>();
            for (Map.Entry<String, String> pattern : RISK_PATTERNS) {
                if (normalized.contains(pattern.getKey()) && !reasons.contains(pattern.getValue())) {
                    reasons.add(pattern.getValue());
                }
            }
            return reasons;
        }

        public boolean isStrictApprovalCommand(String command) {
            String normalized = command == null ? "" : command.trim().toLowerCase(Locale.ROOT);
            for (String pattern : List.of("rm -rf", "rm -fr", "rm -r ", "rm -f /", "mkfs", "dd if=", "> /dev/", "/dev/sd", "/dev/hd", "/dev/nvme", ":(){ :|:& };:")) {
                if (normalized.contains(pattern)) {
                    return true;
                }
            }
            return false;
        }

        public String commandFamily(String command) {
            String trimmed = command == null ? "" : command.trim();
            if (trimmed.isBlank()) {
                return "command";
            }
            String firstToken = trimmed.split("\\s+")[0];
            int slash = Math.max(firstToken.lastIndexOf('/'), firstToken.lastIndexOf('\\'));
            if (slash >= 0) {
                firstToken = firstToken.substring(slash + 1);
            }
            String normalized = firstToken.toLowerCase(Locale.ROOT).trim();
            return normalized.isBlank() ? "command" : normalized;
        }
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/tools/LiteBashTool.kt
package com.zzf.rikki.idea.agent.tools

import com.fasterxml.jackson.databind.JsonNode
import com.zzf.rikki.idea.agent.compat.CommandRunner
import com.zzf.rikki.idea.agent.compat.CommandRunnerResult
import java.util.concurrent.atomic.AtomicBoolean

/** Shell tool facade: validates args, delegates execution, preserves legacy output formatting. */
class LiteBashTool {
    private val runner = CommandRunner()

    fun execute(args: JsonNode, workspaceRoot: String, skipFlag: AtomicBoolean? = null): String {
        val detail = executeDetailed(args, workspaceRoot, skipFlag)
        val timeoutMs = args.path("timeout").asLong(CommandRunner.DEFAULT_TIMEOUT_MS).let {
            if (it <= 0L) CommandRunner.DEFAULT_TIMEOUT_MS else it
        }
        return formatResult(args.path("command").asText(""), timeoutMs, detail)
    }

    fun executeDetailed(
        args: JsonNode,
        workspaceRoot: String,
        skipFlag: AtomicBoolean? = null
    ): CommandRunnerResult {
        val command = args.path("command").asText("").ifBlank {
            throw IllegalArgumentException("command is required")
        }
        val workdir = args.path("workdir").asText(workspaceRoot).ifBlank { workspaceRoot }
        val timeoutMs = args.path("timeout").asLong(CommandRunner.DEFAULT_TIMEOUT_MS).let {
            if (it <= 0L) CommandRunner.DEFAULT_TIMEOUT_MS else it
        }
        val shell = args.path("shell").asText("auto").trim().lowercase().ifBlank { "auto" }
        return runner.run(
            command = command,
            workspaceRoot = workspaceRoot,
            workdir = workdir,
            timeoutMs = timeoutMs,
            shellPref = shell,
            skipFlag = skipFlag
        )
    }

    fun formatResult(command: String, timeoutMs: Long, result: CommandRunnerResult): String =
        runner.formatOutput(command, timeoutMs, result)

    companion object {
        private val RISK_PATTERNS = listOf(
            "sudo " to "requires elevated privileges",
            "su -" to "switches user identity",
            "su root" to "switches user identity",
            "rm -rf" to "destructive file removal",
            "rm -fr" to "destructive file removal",
            "rm -r " to "recursive deletion",
            "rm -f /" to "targets filesystem root",
            "mkfs" to "formats a filesystem",
            "dd if=" to "raw disk write",
            "| bash" to "pipes remote script into shell",
            "| sh " to "pipes remote script into shell",
            "| zsh " to "pipes remote script into shell",
            "| fish " to "pipes remote script into shell",
            "chmod 777" to "grants unsafe world-writable permissions",
            "chmod -R " to "recursively changes permissions",
            "> /dev/" to "writes to device path",
            "/dev/sd" to "targets block device",
            "/dev/hd" to "targets block device",
            "/dev/nvme" to "targets block device",
            ":(){ :|:& };:" to "fork bomb"
        )

        fun isHighRiskCommand(command: String): Boolean = detectRiskReasons(command).isNotEmpty()

        fun detectRiskReasons(command: String): List<String> {
            val normalized = command.trim().lowercase()
            return RISK_PATTERNS
                .filter { normalized.contains(it.first) }
                .map { it.second }
                .distinct()
        }

        fun isStrictApprovalCommand(command: String): Boolean {
            val normalized = command.trim().lowercase()
            return listOf(
                "rm -rf",
                "rm -fr",
                "rm -r ",
                "rm -f /",
                "mkfs",
                "dd if=",
                "> /dev/",
                "/dev/sd",
                "/dev/hd",
                "/dev/nvme",
                ":(){ :|:& };:"
            ).any { normalized.contains(it) }
        }

        fun commandFamily(command: String): String {
            val trimmed = command.trim()
            if (trimmed.isBlank()) {
                return "command"
            }
            val firstToken = trimmed.substringBefore(' ').substringAfterLast('/')
            return firstToken.lowercase().ifBlank { "command" }
        }
    }
}